File: python/update_given_x.py
import numpy as np
from scipy.optimize import minimize
import compute_elbo_bound as compute
import fh
# Define the update_given_x function
def update_given_x(mu, sigma, S, ind_x):
    x = S[:, ind_x].reshape(-1, 1) # OK
    sigma_inv = np.linalg.inv(sigma) # OK, Has same values
    mu_prior = mu.copy().reshape(-1, 1) # OK, Has same values
    prior_muxsigma = (mu.T @ sigma_inv) # Has same values
    w = 1 # OK
    difference = 1 # OK
    iteration_count = 0 # OK

    while difference > 1e-4:
        # Update mean
        mu_update = lambda mu_prior: compute.compute_elbo_bound(mu_prior, sigma_inv, S, sigma, prior_muxsigma, x, w)
        options = {'maxiter': 1e5, 'disp': False}
        res_mu = minimize(mu_update, mu, jac = False, options=options)
        mu = (res_mu.x / np.linalg.norm(res_mu.x)) # Has same values

        # Update covariance
        L0 = np.linalg.cholesky(sigma)
        shape = L0.shape
        c_update = lambda L: fh.fh(L, sigma_inv, S, mu, w, shape) # OK
        res_L = minimize(c_update, L0.flatten(), method='L-BFGS-B', jac = True, options={'maxiter': int(1e5), 'disp': False})
        L = res_L.x.reshape(shape)

        if np.any(np.diag(L0) <= 0): # OK
            temp = np.where(np.diag(L0) <= 0)  # Find indices where diagonal is <= 0
            L[temp, temp] = 1e-14

        sigma = L @ L.T # OK

        difference = np.linalg.norm((mu - mu_prior), ord = 2) # OK
        mu_prior = mu.copy() # OK
        iteration_count += 1 # OK

        if iteration_count > 10: # OK
            break

    return mu, sigma

File: matlab/update_given_x.m
function [mu, sigma] = update_given_x(mu,sigma, S, ind_x)
% UPDATE_GIVEN_X Updates the mean and covariance given the selected word.
%
% Input arguments:
%   mu     - Prior mean vector.
%   sigma  - Prior covariance matrix.
%   S      - Embeddings of words in set.
%   ind_x  - Index of word in set selected by human.
%
% Returns:
%   mu     - Updated mean vector.
%   sigma  - Updated covariance matrix.

    x = S(:, ind_x);
    sigma = double(sigma);
    sigma_inv = inv(sigma);
    mu = double(mu);
    mu_prior = mu;

    % Initialize
    prior_muxsigma = mu' * sigma_inv;
    w = 1;
    difference = 1;
    iteration_count = 0;
    
    while(difference > 1e-4)   
        % Update mean
        mu_update = @(mu_prior) compute_elbo_bound(mu_prior, sigma_inv, S, sigma, prior_muxsigma, x, w);
        options = optimoptions(@fminunc, 'SpecifyObjectiveGradient',false,'MaxIterations', 1e5, 'Display', 'off'); %'HessianFcn' , 'objective' , 
        mu = fminunc(mu_update, mu, options);
        mu = mu./norm(mu);

        % Update covariance
        c_update = @(L) fh(L, sigma_inv, S, mu, w);
        options = optimoptions(@fminunc, 'HessianApproximation' , 'lbfgs' , 'SpecifyObjectiveGradient', true,'MaxIterations', 1e5, 'Display', 'off'); % 'FiniteDifferenceStepSize', 1e-5,
        L0 = chol(sigma, 'lower');
        L = fminunc(c_update, L0, options);
        if(diag(L0)<=0)
            temp = find(diag(L0)<=0);
            L(temp, temp) = 1e-14;
        end
        sigma = L*L';

        % Check termination conditions
        difference = norm(mu-mu_prior);
        mu_prior = mu;
        iteration_count = iteration_count + 1;
        if (iteration_count>10)
            break
        end
    end
end